Beginners course module 3 (#2496)
qdrant/landing_page: qdrant-landing/content/course/beginners/module-3/_index.md

---
title: "Module 3: Sparse vs Dense vs Hybrid Search"
short_description: "Module 3 of the Beginners course: understand dense vs sparse search, when each fails, and how hybrid systems combine them."
description: "Understand dense vs sparse search, when each fails, and how hybrid systems combine them. Learn about fusion strategies, multimodal search, and real-world use cases."
isLesson: true
weight: 40
---

{{< date >}} Module 3 {{< /date >}}

# Sparse vs Dense vs Hybrid Search

> Understand dense vs sparse search, when each fails, and how hybrid systems combine them.

Reviewer: Replace with:

> Understand dense versus sparse retrieval, their strengths, and how a hybrid approach can combine them.

## Today's path

1. Where We Left Off
2. The Two Families of Search
3. Hybrid Search: Dense + Sparse + Filters (Optional)
4. Setting Up Hybrid Search in Qdrant
5. Fusion Strategies
6. Beyond Text: Multimodal Search
7. Real-World Use Cases
8. References & Further Reading

By the end, you'll understand when to use dense, sparse, or hybrid search and how to implement them in Qdrant.

## 1. Where We Left Off

> In Module 2, you built a complete ingestion and retrieval pipeline: raw text → vector → store → top-K query. Dense-only search works well for natural language. It breaks immediately on structured identifiers.

Reviewer: > Dense-only search works well for natural language. It breaks immediately on structured identifiers.

Replace with:

> Dense-only retrieval is best for semantic and contextual search. It struggles on structured identifiers.

| Query | SKU-48291 |
|-------|-----------|
| **The user wants exactly this product. No synonyms. No paraphrasing.** | |

**Dense search returns**

| SKU-48292 | (0.91) | ← wrong |
|----------|--------|---------|
| SKU-48291 | (0.89) | ← correct |

> | SKU-48290 | (0.87) | ← wrong |

Reviewer: Thierry used to do an "iPhone" search example where the wrong model was retrieved for a ecommerce query. I think that works better than SKUs

| DBSF (Distribution-Based Score Fusion) | Normalizes score distributions before merging. Sensitive to relative score differences. | Better when score gaps meaningfully encode relevance and both retrievers are well-calibrated. |

### Starting point

Start with RRF. It's the safer default because dense and sparse scores are on different scales - raw score fusion without normalization produces unreliable results. Switch to DBSF only after evaluating on a labeled test set.

## 6. Beyond Text: Multimodal Search

The same primitive - embed data, store as a vector, search by similarity - applies to any modality. Qdrant stores whatever vectors your embedding model produces. The retrieval mechanics are identical.

🖼️
**Images**
"red dress" → visually similar products
CLIP, SigLIP embed images and text into the same space

🎬
**Video**
"factory fire" → matching video scenes
Frames are sampled, embedded, stored as named vectors

🎧
**Audio**
Hum a melody → matching songs
Audio fingerprints or spectrogram embeddings

📝
**Text**
"cheap flights NYC" → semantic docs

> Sentence transformers, OpenAI embeddings, etc.

Reviewer: I don't think we want emojis

  - Live comparison of BM25, SPLADE, and dense embeddings on the same queries.

- **Hybrid Queries (RRF + DBSF)** - [Hybrid Queries - Qdrant](https://qdrant.tech/documentation/concepts/hybrid_query/)
  - Full reference for prefetch, FusionQuery, RRF, and DBSF in the Qdrant API.

- **Sparse Vectors Reference** - [Vectors - Qdrant](https://qdrant.tech/documentation/concepts/vectors/#sparse-vectors)
  - SparseVectorParams, index configuration, and storage options.

- **Multimodal Search Tutorial** - [Multimodal and Multilingual RAG with LlamaIndex and Qdrant](https://qdrant.tech/documentation/tutorials/multimodal_rag/)
  - Text + image search with embeddings and named vectors.

- **Named Vectors** - [Collections - Qdrant](https://qdrant.tech/documentation/concepts/collections/#multiple-vectors)
  - How to configure and query multiple named vectors on the same point.

## What's Next - Module 4

Next, we'll explore:

- Real enterprise architectures - Tripadvisor, HubSpot, and OpenTable in depth
- Production patterns: multi-tenancy, agent memory, and RAG pipelines
- Deployment options: Cloud, Hybrid Cloud, Edge, and self-hosted

> - Formula queries - when RRF and DBSF aren't enough

Reviewer: Do not use real customer use cases in the course

isLesson: true
weight: 40
---

{{< date >}} Module 3 {{< /date >}}

# Sparse vs Dense vs Hybrid Search

<div class="video">
  <iframe src="https://www.youtube.com/embed/9XXz21jmWes?rel=0" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen>
  </iframe>
</div>

Understand dense versus sparse retrieval, their strengths, and how a hybrid approach can combine them.

**Follow-along code**: [Module 3 notebook](https://github.com/qdrant/examples/blob/master/course/beginners/Module3.ipynb)

#### TL;DR
```
Module 2 showed you where your data lives and how Qdrant retrieves it. 
In this module, you'll learn what that retrieval misses and how to cover the gap. 
You'll explore dense and sparse vectors, BM25, and the inverted index, then see 
why a product code defeats either one alone. You'll also learn how fusion merges 
two ranked lists, and where a filter belongs so both retrievers respect it. 
By the end, you'll have built a hybrid collection, run a fused query, 
and filtered it correctly.

> ```

Reviewer 2: ```suggestion
Module 2 showed you where your data lives and how Qdrant retrieves it. 
In this module, you'll learn what that retrieval misses and how to cover the gap. 
You'll explore dense and sparse vectors, BM25, and the inverted index, then see 
why a product code defeats either one alone. You'll also learn how fusion merges 
two ranked lists, and where a filter belongs so both retrievers respect it. 
By the end, you'll have built a hybrid collection, run a fused query, 
and filtered it correctly.
```
Remove code quotation marks

cosine("cheap flights", "bake a cake")              # 0.2047
```

Read the gap, not the absolute number: what makes 0.73 meaningful is the distance from the 0.20, not the value on its own.

### Sparse Search

![A sparse vector drawn as a long mostly-empty row, with weights on only the handful of positions matching tokens present in the text.](/courses/beginners/module-3/sparse-search.png)

Sparse vectors are token-based: each dimension corresponds to a token, only the tokens present carry a weight, and everything else is zero. Storing that mostly-zero row would be wasteful, so a sparse vector is two parallel arrays, the `indices` of the non-zero dimensions and the `values` at those positions:

```python
# BM25 vector for "Nike Pegasus 40 running shoes"
indices = [1974139272, 24614856, 1784631546, 243905464, 303109060]
values  = [1.67, 1.67, 1.67, 1.67, 1.67]
```

Five tokens, five weights, nothing else stored. The indices are hashes of each token rather than positions in a word list, which is why they are large. The values are uniform because BM25 produces only half the score, the part counting how often a token appears. Qdrant computes the other half at query time.

> Sparse retrieval does not match characters, which is a common assumption. BM25 tokenizes and stems first, so `SKU-48291` and `SKU-48292` still share the token `sku`. What it gives you is that `40` and `41` are *different tokens* with no relationship at all, where dense placed them 0.0087 apart. The distinguishing token gets its own dimension instead of being averaged away.

Reviewer 2: ```suggestion
Sparse retrieval does not match characters, which is a common assumption. M25 first splits text into tokens and cuts each token back to its root. (stemming), so `SKU-48291` and `SKU-48292` still share the token `sku`. <br>
What it gives you is that `40` and `41` are *different tokens* with no relationship at all, where dense placed them 0.0087 apart. The distinguishing token gets its own dimension instead of being averaged away.
```
define stemming

Sparse similarity in Qdrant is always the dot product, with no metric to choose, unlike the dense side where you pick Cosine, Dot, or Euclidean.

#### Sparse Models

> This module uses **BM25**: statistical, no training needed, and it weights the tokens a text contains without adding any it does not. Two neural alternatives go further. **SPLADE** expands a text with related terms, and **miniCOIL** keeps BM25's vocabulary but weights each token using its context, which is what Qdrant recommends for new projects. Swapping between them is a one-line model change, so start here and read [Understanding SPLADE and Sparse Vectors](/articles/sparse-vectors/) or [miniCOIL](/articles/minicoil/) when you need more.

Reviewer 2: ```suggestion
This module uses BM25, a statistical scoring method that requires no training. It scores the terms already present in the text.

Two trained models extend BM25:

- **SPLADE** expands the text with related terms, allowing a document to match a query even when they share none of the same words.
- **miniCOIL** keeps the original terms but weights them based on their surrounding context. We recommend it for new projects.

Switching between these models requires only a one-line change. Start with BM25, then read [Understanding SPLADE and Sparse Vectors](/articles/sparse-vectors/) or [miniCOIL](/articles/minicoil/) when you need more.
```
Break down the paragraph

#### Indexing Sparse Vectors

> Because most dimensions are zero, Qdrant stores sparse vectors in an **inverted index**: rather than a row per point, it keeps one list per token recording which points contain that token and with what weight. Each list is a **posting list**.

Reviewer 2: ```suggestion
Because most dimensions are zero, Qdrant stores sparse vectors in an inverted index: rather than a row per point, it keeps one list per token recording which points contain that token and its weight.
```
Posting list is not a term we use anywhere

```text
Token "nike"    → posting list: [point_1, point_2, point_4, ...]
Token "pegasus" → posting list: [point_1, point_2, point_3, ...]
```

> A query reads only the posting lists for the tokens it contains, summing weights as it goes, and never touches a point sharing none of them. HNSW from Module 2 is approximate, trading a little accuracy for speed. The sparse index makes no such trade: it is exact.

Reviewer 2: ```suggestion
A query reads only the lists for the tokens it contains, summing weights as it goes, so it scores just the points that share at least one token with the query. HNSW from Module 2 is approximate, trading a little accuracy for speed. The sparse index scores every candidate those lists hold, so its ranking is exact.
```
Remove posting list

### Dense and Sparse, Side by Side

> Module 1 covered the strengths and they hold here. Dense handles synonyms, paraphrasing, and intent, and struggles wherever one distinguishing token carries the meaning: serial numbers, rare words, invented product names. Sparse is the mirror image, strong on exact tokens and domain jargon, and blind to any rewording that shares no words. One caveat on cross-language retrieval: the model used here, `sentence-transformers/all-MiniLM-L6-v2`, is English-only, so it needs the multilingual model Module 1 discussed.

Reviewer 2: ```suggestion
Module 1 covered the strengths of dense vectors, and those strengths apply here too. Dense vectors handle synonyms, paraphrases, and intent well, but they can miss cases where one exact token carries the meaning, such as a serial number, rare term, or invented product name. <br>
Sparse vectors make the opposite tradeoff. They work well for exact tokens and domain-specific terms, but they cannot recognize reworded content when the query and document share no words. <br>
One caveat is cross-language retrieval. The model used here, `sentence-transformers/all-MiniLM-L6-v2`, supports English only. For cross-language retrieval, use a multilingual embedding model.
```
Break down the paragraph

> Dense finds what a query means. Sparse finds what it says. Product search, support search, and code search usually need both at once, which is what the next section is for.

Reviewer 2: ```suggestion
Dense finds what a query means. Sparse finds what it says. Most real catalogs carry both, a name a shopper paraphrases and a model number they type exactly, which is what the next section builds.
```
Better to use a concrete example

## 3. Hybrid Search: Dense and Sparse

Hybrid search runs both retrievers in the same request and combines their ranked lists into one result set.

Run the same `Nike Pegasus 40` query through sparse alone and the picture inverts. The top three:

| Result | Sparse score |
|--------|--------------|
| Nike Pegasus 40 running shoes | 3.8396 |
| Nike Pegasus 40 womens running shoes | 3.8293 |
| Nike Pegasus 41 running shoes | 1.7007 |

Sparse pushes the 41 down to third, because `40` is a different token from `41`. But it now has dense's problem on a different pair: the men's and women's Pegasus 40 share every token the query contains, so it separates them by 0.0103 on a 3.8396 top score, roughly a quarter of a percent. Dense had those two 0.09 apart and no trouble at all.

Each retriever ranks the right shoe first, and each leaves it a hair ahead of something wrong. Neither is safe alone.

![Hybrid search for the query "Nike Pegasus 40": dense retrieval contributes semantically related running shoes while sparse retrieval locks onto the exact model number, and fusion combines both into one ranked list.](/courses/beginners/module-3/nike-example.png)

### Fusion

> **Fusion** is the step that merges two ranked lists into one. It runs on the server once both retrievers finish, and it decides the final order. Qdrant offers two strategies, Reciprocal Rank Fusion and Distribution-Based Score Fusion, and the [Hybrid Queries documentation](/documentation/search/hybrid-queries/) covers their parameters.

Reviewer 2: ```suggestion
**Fusion** combines the ranked results from two retrievers into a single list. After both retrievers finish, Qdrant applies fusion on the server to determine the final ranking. <br>
Qdrant supports two fusion strategies: Reciprocal Rank Fusion and Distribution-Based Score Fusion. See the [Hybrid Queries documentation](/documentation/search/hybrid-queries/) for details and available parameters.
```
Not a fan of and, and

Install the client with FastEmbed support:

```bash
pip install "qdrant-client[fastembed]"
```

Then create the collection and its payload indexes:

```python
from qdrant_client import QdrantClient, models

client = QdrantClient(
    url="https://YOUR-CLUSTER.cloud.qdrant.io",
    api_key="YOUR_API_KEY",
)

client.create_collection(
    collection_name="products",
    vectors_config={
        "dense": models.VectorParams(size=384, distance=models.Distance.COSINE),
    },
    sparse_vectors_config={
        "sparse": models.SparseVectorParams(
            modifier=models.Modifier.IDF,
        ),
    },
)

client.create_payload_index(
    collection_name="products",
    field_name="in_stock",
    field_schema=models.PayloadSchemaType.BOOL,
)
client.create_payload_index(
    collection_name="products",
    field_name="sizes",
    field_schema=models.PayloadSchemaType.INTEGER,
)
client.create_payload_index(
    collection_name="products",
    field_name="price",
    field_schema=models.PayloadSchemaType.FLOAT,
)
```

> Index before ingestion. An index added later still filters correctly, but the structure Qdrant uses to skip non-matching points during a search is built as the data is written, so adding one afterward means a rebuild. Qdrant Cloud also runs **strict mode** by default, a set of guardrails that rejects queries expensive enough to destabilize a cluster, and filtering an unindexed field is one of them: you get a 400 rather than a slow answer.

Reviewer 2: ```suggestion
**Index Before Ingestion.** Create payload indexes before you ingest data. Qdrant can add an index later and still filter correctly, but it must rebuild the index for existing points. Creating it first lets Qdrant build the index as it writes the data. <br>
Qdrant Cloud also enables **strict mode** by default. These guardrails reject queries that could be expensive enough to destabilize a cluster. Filtering on an unindexed field is one such case, so Qdrant returns a `400` error instead of running a slow query.
```
More clear

0.2000  Brooks Ghost 15 neutral running shoes
```

The women's Pegasus 40 was second in the unfiltered run and is gone: out of stock, and it stops at size 8. The Trail 4 is gone too, since it stops at size 10. Neither was retrieved and discarded; they never entered a candidate list.

### Filtering a Dense or Sparse Query

With no prefetch, the filter belongs at the top level as `query_filter`:

```python
results = client.query_points(
    collection_name="products",
    query=models.Document(text="Nike Pegasus 40", model=DENSE_MODEL),
    using="dense",
    query_filter=shopper_filter,
    limit=4,
).points
```

Swap `DENSE_MODEL` for `SPARSE_MODEL` and `using="sparse"` for the sparse equivalent. The placement is identical because there is no prefetch for the filter to belong to.

Whichever retriever you use, the filter admits the same four products, the ones in stock and stocked in size 11: the Pegasus 40 and 41, the Invincible 3, and the Ghost 15. Eligibility does not depend on the retriever. Only the order they come back in does, which is the whole reason the choice of retriever still matters after filtering.

### One Thing to Watch: Local Mode

> A hybrid query accepts a filter in either place, and a real server honors both: a top-level `query_filter` does reach the prefetches, so results stay valid either way. Local mode is where they diverge. It ignores the top-level filter and raises no error, so a notebook can print results that break their own filter while looking healthy. Per-prefetch placement behaves the same everywhere, which is reason enough to make it the habit.

Reviewer 2: ```suggestion
<aside role="status">
The companion notebook uses local mode with `QdrantClient(":memory:")`. In this mode, Qdrant ignores a top-level filter on a hybrid query instead of returning an error, so the notebook may show results that don’t match the filter.

Filters defined inside each `Prefetch` work consistently in both local mode and Qdrant Cloud.
</aside>
```
More clear

2. Add a `models.FieldCondition` on `price` with `models.Range(lte=140)` to `shopper_filter` and rerun. Two products should drop out. Predict which before you run it.
3. Query `comfortable shoes for long runs`, a phrase no title contains. Dense ranks the products, while sparse returns a flat tie across the five titles holding both `running` and `shoes`. Work out why sharing exactly the same query tokens produces a tie.

## 7. Beyond Text: Multimodal Search

The same primitive, embed data and search by similarity, applies to any modality: images with CLIP or SigLIP, video as sampled frames, audio as spectrogram embeddings. Qdrant stores whatever vectors your model produces and the retrieval mechanics do not change.

Store them as named vectors on the same point, exactly as this module stored dense and sparse together. Each named vector is its own space, so a query has to be embedded by the model that produced the vectors it is searching: to find images with a text query, embed that text with CLIP's text encoder, not the sentence transformer. Module 5 builds this out as the capstone.

> ## 8. Knowledge Check

Reviewer 2: Hide the answers in `<details>` blocks. Showing them immediately turns the exercise into a recap and removes the need to think.

Ask readers to make a decision, not repeat a sentence from the previous section. Three of the four questions currently test recall. Keep questions that place readers in scenarios where they must apply the module’s concepts:

## 9. References & Further Reading

- [Hybrid Queries](/documentation/search/hybrid-queries/): prefetch semantics, both fusion strategies, and their tuning parameters.
- [Understanding SPLADE and Sparse Vectors](/articles/sparse-vectors/): how sparse vectors work and how SPLADE compares to BM25.
- [miniCOIL: Sparse Neural Retrieval](/articles/minicoil/): why miniCOIL exists and how it extends BM25 with contextual meaning.
- [Filtering](/documentation/search/filtering/): full filter syntax and the payload index each condition needs.
- [Named Vectors](/documentation/manage-data/vectors/#named-vectors): configuring and querying more than one vector on a single point.

## What's Next - Module 4

Next we go from primitives to judgment, designing a complete vector search system end to end:

- The five layers of a vector search stack: query, indexing, storage, knowledge, and distribution
- A worked example: designing a news search system, decision by decision
- Filtering in production: how the query planner combines filters with vector search, and multitenancy via payload filters
- The production RAG pipeline, from query understanding to generation

> - Deployment options: Local, Docker, Managed Cloud, Hybrid Cloud, Private Cloud, and Edge

Reviewer 2: ```suggestion
Eight products can fit in one collection and one query. A multilingual news archive with strict tenant isolation and a generation step requires a more deliberate design. <br>
Module 4 works through this system end to end, decision by decision. You’ll learn how the layers of the stack work together, how filters behave when the query planner selects a strategy, and how deployment choices range from Docker to Qdrant Edge.
```
Better hook, no need to give the full details

---
title: "Module 3: Sparse vs Dense vs Hybrid Search"
short_description: "Module 3 of the Beginners course: dense and sparse retrieval, what each one misses, and how hybrid search combines them."
description: "Compare dense and sparse retrieval, see where each one fails on real queries, and build a hybrid search pipeline in Qdrant with rank fusion and filters."
isLesson: true
weight: 40
---

{{< date >}} Module 3 {{< /date >}}

# Sparse vs Dense vs Hybrid Search

<div class="video">

>   <iframe src="https://www.youtube.com/embed/9XXz21jmWes?rel=0" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen>

Reviewer: video needs updating with final, branded version that is audio-leveled

  </iframe>
</div>

Understand dense versus sparse retrieval, their strengths, and how a hybrid approach can combine them.

**Follow-along code**: [Module 3 notebook](https://github.com/qdrant/examples/blob/master/course/beginners/Module3.ipynb)

> #### TL;DR

Reviewer: This is not a TLDR, its an Overview. Either content or section header needs to change (and be consistent across modules)

> Module 2 showed you where your data lives and how Qdrant retrieves it. 
In this module, you'll learn what that retrieval misses and how to cover the gap. 
You'll explore dense and sparse vectors, BM25, and the inverted index, then see 
why a product code defeats either one alone. You'll also learn how fusion merges 
two ranked lists, and where a filter belongs so both retrievers respect it. 
By the end, you'll have built a hybrid collection, run a fused query, 

> and filtered it correctly.

Reviewer: Not sure why all of this is on separate lines

## Today's Path

1. Where We Left Off
2. The Two Families of Search
3. Hybrid Search: Dense and Sparse
4. Setting Up Hybrid Search in Qdrant
5. Fusion Strategies
6. Filtering: Works with Any Retrieval Method
7. Knowledge Check
8. References & Further Reading

## 1. Where We Left Off

> In Module 2 you built a complete ingestion and retrieval pipeline: raw text, vector, store, top-K query. Dense retrieval handles meaning well. It gets shaky on the part of a query that has to be exact.

Reviewer: add comma after "Module 2"

Here is a shoe catalog with two products one digit apart. Searching it dense-only for `Nike Pegasus 40`:

| Result | Dense score |
|--------|-------------|
| Nike Pegasus 40 running shoes | 0.8713 |
| Nike Pegasus 41 running shoes | 0.8626 |
| Nike Pegasus 40 womens running shoes | 0.7830 |
| Nike Pegasus Trail 4 trail running shoes | 0.7425 |

### The Problem

Dense ranks the right shoe first, so nothing here looks broken. Look at the margin: **0.0087**, about one percent of the top score. To the model, "Pegasus 40" and "Pegasus 41" are near-identical statements about running shoes, because that is what they are. The digit a shopper cares about is one token out of five, averaged into a vector describing the whole phrase.

A margin that thin holds across eight products. Across eighty thousand, with every colorway and width in the catalog, something will drift into that one percent and take the top slot. Model numbers, SKUs, and part codes need matching, not neighborhood. That is the gap sparse search fills.

## 2. The Two Families of Search

Every retrieval system is built from one or both of these.

### Dense Search

> ![Two short phrases encoded into 384-dimensional dense vectors, positioned close together in vector space because they share meaning rather than words.](/courses/beginners/module-3/dense-search.png)

Reviewer: remove title from diagram and fix branding

A dense vector has a small, fixed number of dimensions, 384 for the model used here, and every one of them holds a value. Two texts with similar meaning land close together whether or not they share any words:

```python
cosine("car repair",    "automobile maintenance")   # 0.7334
cosine("cheap flights", "affordable airfare")       # 0.7241
cosine("cheap flights", "bake a cake")              # 0.2047
```

Read the gap, not the absolute number: what makes 0.73 meaningful is the distance from the 0.20, not the value on its own.

### Sparse Search

> ![A sparse vector drawn as a long mostly-empty row, with weights on only the handful of positions matching tokens present in the text.](/courses/beginners/module-3/sparse-search.png)

Reviewer: remove title from diagram and fix branding

```

Then create the collection and its payload indexes:

```python
from qdrant_client import QdrantClient, models

client = QdrantClient(
    url="https://YOUR-CLUSTER.cloud.qdrant.io",
    api_key="YOUR_API_KEY",
)

client.create_collection(
    collection_name="products",
    vectors_config={
        "dense": models.VectorParams(size=384, distance=models.Distance.COSINE),
    },
    sparse_vectors_config={
        "sparse": models.SparseVectorParams(
            modifier=models.Modifier.IDF,
        ),
    },
)

client.create_payload_index(
    collection_name="products",
    field_name="in_stock",
    field_schema=models.PayloadSchemaType.BOOL,
)
client.create_payload_index(
    collection_name="products",
    field_name="sizes",
    field_schema=models.PayloadSchemaType.INTEGER,
)
client.create_payload_index(
    collection_name="products",
    field_name="price",
    field_schema=models.PayloadSchemaType.FLOAT,
)
```

**Index Before Ingestion.** Create payload indexes before you ingest data. Qdrant can add an index later and still filter correctly, but it must rebuild the index for existing points. Creating it first lets Qdrant build the index as it writes the data. <br>

> Qdrant Cloud also enables **strict mode** by default. These guardrails reject queries that could be expensive enough to destabilize a cluster. Filtering on an unindexed field is one such case, so Qdrant returns a `400` error instead of running a slow query.

Reviewer: link to strict mode docs

                limit=20,
            ),
            models.Prefetch(
                query=models.Document(text=query_text, model=SPARSE_MODEL),
                using="sparse",
                limit=20,
            ),
        ],
        # Fusion merges the two candidate lists by rank position
        query=models.RrfQuery(rrf=models.Rrf()),
        limit=limit,
    ).points

for r in hybrid_search("Nike Pegasus 40"):
    print(f"{r.score:.4f}  {r.payload['title']}")
```

Real output:

```text
1.0000  Nike Pegasus 40 running shoes
0.5833  Nike Pegasus 41 running shoes
0.5833  Nike Pegasus 40 womens running shoes
0.4000  Nike Pegasus Trail 4 trail running shoes
```

Every run in this module put the right shoe first, so the order is not what improved. What changed is the margin: dense and sparse alone each left the right shoe under two percent clear of a rival (Sections 1 and 3), while hybrid with RRF widens that lead to **41.7%** (1.0000 vs. 0.5833). The Pegasus 40 is the only product ranked first by *both* retrievers, and fusion turns that agreement into distance. Fusion scores come from rank position, which is why they resemble neither input scale.

## 5. Fusion Strategies

> 

Reviewer: section needs a diagram
